platform/jewel/ui/src/main/kotlin/org/jetbrains/jewel/ui/component/popup/JDialogRenderer.kt

```diff
@@ -0,0 +1,452 @@
+                content = content,
+            )
+
+            return
+        }
+
+        JPopupImpl(
+            window = window,
+            popupPositionProvider = popupPositionProvider,
+            onDismissRequest = onDismissRequest,
+            properties = properties,
+            onPreviewKeyEvent = onPreviewKeyEvent,
+            onKeyEvent = onKeyEvent,
+            cornerSize = cornerSize,
+            blendingEnabled = supportBlending,
+            content = content,
+        )
+    }
+}
+
+@Composable
+private fun JPopupImpl(
+    window: Window,
+    popupPositionProvider: PopupPositionProvider,
+    onDismissRequest: (() -> Unit)?,
+    properties: PopupProperties,
+    onPreviewKeyEvent: ((KeyEvent) -> Boolean)?,
+    onKeyEvent: ((KeyEvent) -> Boolean)?,
+    cornerSize: CornerSize,
+    blendingEnabled: Boolean,
+    content: @Composable () -> Unit,
+) {
+    val popupDensity = LocalDensity.current
+    val component = LocalComponent.current
+
+    val currentContent = rememberUpdatedState(content)
+    val currentPopupPositionProvider = rememberUpdatedState(popupPositionProvider)
+
+    val compositionLocalContext = currentCompositionLocalContext
+
+    val parentBoundsInRoot = remember { mutableStateOf<IntRect?>(null) }
+    val popupRectangle = remember { mutableStateOf<Rectangle?>(null) }
+
+    Layout(
+        content = {},
+        modifier =
+            Modifier.onGloballyPositioned { childCoordinates ->
+                childCoordinates.parentCoordinates?.let {
+                    parentBoundsInRoot.value = it.boundsInRoot().roundToIntRect().fromRelativeToScreen(component)
+                }
+            },
+        measurePolicy = { _, _ -> layout(0, 0) {} },
+    )
+
+    val currentOnDismissRequest = rememberUpdatedState(onDismissRequest)
+    val currentOnKeyEvent = rememberUpdatedState(onKeyEvent)
+
+    val overriddenOnKeyEvent =
+        if (properties.dismissOnBackPress && onDismissRequest != null) {
+            // No need to remember this lambda, as it doesn't capture any values that can change.
+            { event: KeyEvent ->
+                val consumed = currentOnKeyEvent.value?.invoke(event) ?: false
+                if (!consumed && event.isDismissRequest()) {
+                    currentOnDismissRequest.value?.invoke()
+                    true
+                } else {
+                    consumed
+                }
+            }
+        } else {
+            onKeyEvent
+        }
+
+    val dialog = remember {
+        JDialog(window).apply {
+            isAlwaysOnTop = true
+            isFocusable = properties.focusable
+            focusableWindowState = properties.focusable
+            isUndecorated = true
+            rootPane.isOpaque = false
+            background = Color(0, 0, 0, 0)
+            contentPane.background = Color(0, 0, 0, 0)
+            rootPane.putClientProperty("Window.shadow", true)
+        }
+    }
+
+    val composePanel = remember {
```

Comment:
This remember computation also accesses a lot of things that may change, do they need to be keys? I'm not sure how it works with nested compositions

(I realise a lot of this may also affect the JBPopup implementation... I did not pick them up at the time)

Reply:
That's something I was thinking, but since all the access I do is to "state" types, we shouldn't need to recreate the compose panel; instead, they should "recompose" as expected.